platform/posix/transport/src/openssl_posix.c

```diff
@@ -0,0 +1,674 @@
+ *
+ * @param[out] pSslSetup Destination for the imported credentials.
+ * @param[in] pOpensslCredentials TLS credentials to be validated.
+ *
+ * @return 1 on success; otherwise, failure;
+ */
+static int setCredentials( SSL_CTX * pSslSetup,
+                           const OpensslCredentials_t * pOpensslCredentials );
+
+/**
+ * @brief Set optional configurations for the TLS connection.
+ *
+ * This function is used to set SNI, MFLN, and ALPN protocols.
+ *
+ * @param[in] pSslContext Destination for the imported credentials.
+ * @param[in] pOpensslCredentials TLS credentials containing configurations.
+ */
+static void setOptionalConfigurations( SSL * pSslContext,
+                                       const OpensslCredentials_t * pOpensslCredentials );
+
+/*-----------------------------------------------------------*/
+
+static void logPath( const char * path,
+                     size_t pathLen,
+                     const char * fileType )
+{
+    char * cwd = getcwd( NULL, 0 );
+
+    /* Log the absolute directory based on first character of path. */
+    if( ( path[ 0 ] == '/' ) || ( path[ 0 ] == '\\' ) )
+    {
+        LogDebug( ( "Attempting to open %s: Path=%.*s.",
+                    fileType,
+                    ( int32_t ) pathLen,
+                    path ) );
+    }
+    else
+    {
+        LogDebug( ( "Attempting to open %s: Path=%s/%.*s.",
+                    fileType,
+                    cwd,
+                    ( int32_t ) pathLen,
+                    path ) );
+    }
+
+    /* Free cwd because getcwd calls malloc. */
+    if( cwd != NULL )
+    {
+        free( cwd );
+    }
+}
+
+static int setRootCa( SSL_CTX * pSslSetup,
+                      const char * pRootCaPath,
+                      size_t rootCaPathLen )
+{
+    int sslStatus = 0;
+    FILE * pRootCaFile = NULL;
+    X509 * pRootCa = NULL;
+
+    logPath( pRootCaPath, rootCaPathLen, ROOT_CA_LABEL );
+
+    pRootCaFile = fopen( pRootCaPath, "r" );
+
+    if( pRootCaFile == NULL )
+    {
+        LogError( ( "fopen failed to find the root CA certificate file: "
+                    "ROOT_CA_PATH=%.*s.",
+                    ( int32_t ) rootCaPathLen,
+                    pRootCaPath ) );
+        sslStatus = -1;
+    }
+    else
+    {
+        /* Read the root CA into an X509 object, then close its file handle. */
+        pRootCa = PEM_read_X509( pRootCaFile, NULL, NULL, NULL );
+
+        if( fclose( pRootCaFile ) != 0 )
+        {
+            LogWarn( ( "fclose failed to close file %.*s",
+                       ( int32_t ) rootCaPathLen,
+                       pRootCaPath ) );
+        }
+    }
+
+    if( pRootCa == NULL )
+    {
+        LogError( ( "PEM_read_X509 failed to parse root CA." ) );
+        sslStatus = -1;
+    }
+    else
+    {
+        sslStatus = X509_STORE_add_cert( SSL_CTX_get_cert_store( pSslSetup ),
+                                         pRootCa );
+    }
+
+    if( sslStatus != 1 )
```

Comment:
This should be in the above else block, otherwise it will be printed even when `pRootCa = NULL`